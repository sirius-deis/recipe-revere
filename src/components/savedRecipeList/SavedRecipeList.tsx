import { FC } from "react";
import { useQuery } from "@apollo/client";
import Loader from "../loader/Loader";
import SavedRecipe from "../savedRecipe/SavedRecipe";
import List from "../list/List";
import { GET_SAVED_RECIPES } from "../../queries/queries";
import ErrorBox from "../errorBox/ErrorBox";

const SavedRecipeList = () => {
  const { loading, error, data } = useQuery(GET_SAVED_RECIPES);
  if (loading) {
    return <Loader />;
  }
  if (error) {
    return <ErrorBox message={error.message} />;
  }
  return (
    <section>
      <List direction="horizontal">
        {data.recipes.map(
          (recipe: { url: string; label: string; image: string }) => (
            <SavedRecipe
              key={recipe.url}
              imgUrl={recipe.image}
              title={recipe.label}
              _id={recipe.url}
            />
          )
        )}
      </List>
    </section>
  );
};

export default SavedRecipeList;
