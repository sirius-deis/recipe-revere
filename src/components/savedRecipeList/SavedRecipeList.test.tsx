import { render, screen } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import SavedRecipeList from "./SavedRecipeList";

jest.mock("@apollo/client", () => ({
  ...jest.requireActual("@apollo/client"),
  useQuery: jest.fn(),
}));
import { useQuery } from "@apollo/client";

jest.mock("../loader/Loader", () => () => <div>Loading...</div>);

describe("SavedRecipeList", () => {
  const mockedData = {
    recipes: [
      { url: "1", label: "Recipe 1", image: "img1.jpg" },
      { url: "2", label: "Recipe 2", image: "img2.jpg" },
    ],
  }
  it("should match snapshot", () => {
    (useQuery as jest.Mock).mockReturnValue({
      loading: false,
      error: null,
      data: mockedData,
    });
    const { container } = render(
      <Router>
        <SavedRecipeList />
      </Router>
    );
    expect(container).toMatchSnapshot();
  })
  it("should render loading state", () => {
    (useQuery as jest.Mock).mockReturnValue({
      loading: true
    });
    render(<Router>
      <SavedRecipeList />
    </Router>);
    expect(screen.getByText("Loading...")).toBeInTheDocument();
  })
  it("should render error state", () => {
    const errorMessage = "Error fetching saved recipes";
    (useQuery as jest.Mock).mockReturnValue({
      loading: false,
      error: new Error(errorMessage),
    });
    render(<Router>
      <SavedRecipeList />
    </Router>);
    expect(screen.getByText(errorMessage)).toBeInTheDocument();
  });
  it("should render no saved recipes message", () => {
    (useQuery as jest.Mock).mockReturnValue({
      loading: false,
      error: null,
      data: { recipes: [] }
    })
    render(<Router>
      <SavedRecipeList /></Router>);
    expect(screen.getByText("No saved recipes yet")).toBeInTheDocument();
  })
  it("should render saved recipes", () => {
    (useQuery as jest.Mock).mockReturnValue({
      loading: false,
      error: null,
      data: mockedData
    })
    render(<Router>
      <SavedRecipeList />
    </Router>)
    expect(screen.queryByTitle("Loading...")).not.toBeInTheDocument();
    for (let recipe of mockedData.recipes) {
      expect(screen.getByText(recipe.label)).toBeInTheDocument();
      const img = screen.getByAltText(recipe.label);
      expect(img).toBeInTheDocument();
      expect(img).toHaveAttribute('src', recipe.image)
    }
  })
})