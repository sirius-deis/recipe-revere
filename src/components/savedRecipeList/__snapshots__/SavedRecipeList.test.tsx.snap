// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SavedRecipeList should match snapshot 1`] = `
<div>
  <section>
    <ul
      class="list horizontal"
    >
      <li
        class="list-item"
      >
        <div
          class="panel shadowed centered vertical"
          data-testid="panel"
        >
          <div
            class="image-container"
          >
            <img
              alt="Recipe 1"
              class="img"
              src="img1.jpg"
            />
          </div>
          <h3
            class="title"
          >
            Recipe 1
          </h3>
          <a
            class="link"
            href="/recipe/1"
            rel="noreferrer"
            target="_blank"
          >
            Details
          </a>
        </div>
      </li>
      <li
        class="list-item"
      >
        <div
          class="panel shadowed centered vertical"
          data-testid="panel"
        >
          <div
            class="image-container"
          >
            <img
              alt="Recipe 2"
              class="img"
              src="img2.jpg"
            />
          </div>
          <h3
            class="title"
          >
            Recipe 2
          </h3>
          <a
            class="link"
            href="/recipe/2"
            rel="noreferrer"
            target="_blank"
          >
            Details
          </a>
        </div>
      </li>
    </ul>
  </section>
</div>
`;
